Optimize selection of part list during query planning

jawm opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Up to 8x faster `SELECT` queries with heavy partition pruning on tables with 10K+ parts

### Documentation entry for user-facing changes

Benchmark - https://pastila.nl/?003033a7/21217bba3c0e5174038bb4f998d6c930#HxLYLrj0MFz2a283lYl/Ug==

### Implementation details

- Split data parts lock into read and write
- Introduce shared version of active parts and parts ranges for `SELECT` queries, that is updated on each `SELECT` if required (each acquire of parts lock for write reset them)

### Refs
- https://github.com/ClickHouse/ClickHouse/pull/86933

### Details

We (Cloudflare) initially built out this change on an older checkout of ClickHouse, to help resolve a scaling issue in one of our clusters. This cluster has tables with quite a high part count, and a high level of select concurrency. Over time, we observed an increasing amount of time being spent in query planning, eventually more than half of the query duration.

This was initially observed to be due to lock contention in MergeTreeData, so I made a change that read-only functions could use a shared lock instead of a unique lock. After this it was still visible that a surprising amount of time was being spent copying the parts vector, so I've also made a shared copy of it, which is shared amongst all SELECT queries, until the set of parts is modified, at which point the cached list gets recomputed.

Overall these changes have significantly improved the performance in this cluster, reducing leaf query duration by around two thirds. We recognise that the table was probably not set up entirely in alignment with recommended guidelines, but I think it's still worthwhile applying this change, as it overall only brought improvements for us. In more conventional tables, which do follow guidelines, we observed no negative impact from these changes.

The PR is formed of two commits, which I cherry-picked from our internal checkout, with some conflict-resolution needed too. Each of those brings a separate optimisation as described above, and they can be reviewed individually, although the second commit builds on the first, so it does make sense to review them in order.

jawm committed:
Cache a shared vector of parts in MergeTreeData.

Even using a shared lock when acquiring the vector of parts is limited
in value, since I still observe quite a bit of time spent copying the
parts vector in the `getDataPartsVectorForInternalUsage` function. Given
this list of parts is already getting filtered later in the query
planning phase, it seems that sharing a copy of the vector across all
concurrently running SELECT queries makes sense.

The implementation of this caching depends on the previous commit, which
brings the option to either "read lock" or "write lock" the merge tree
data. In the case of a write lock, we know that this implies the set of
parts may be getting mutated, and as such it would invalidate our cached
list of parts. So, in the destructor for this lock, we recompute the
cached list.

This causes the "read vs write" lock to have a more important
difference, as a "write lock" is no longer const, and as such cannot be
called from const functions. This is why quite a few callsites needed
modified in this commit.

This brings a significant improvement to query planning performance,
particularly for tables with many parts, or where many queries are
concurrently accessing a table.

jawm committed:
Allow taking a shared lock of parts in MergeTreeData

In one of our internal clusters (at Cloudflare), we encountered a
situation where the majority of leaf query duration was spent waiting to
acquire the parts lock. In particular when calling
`MergeTreeData::getStorageSnapshot`.

This particular cluster has a table design which leads to quite a high
part count, which is problematic, as some paths which hold the parts
lock have to iterate over all paths. Eventually acquiring the lock
becomes more of a bottleneck than the actual work being done once the
lock is acquired.

My solution was to make it so that read-only operations over the parts
list can use a shared lock, allowing multiple entrants to hold the lock
simultaneously. This completely resolved the contention problem for us,
and *significantly* improved query performance for this workload,
without showing any negative impact to more conventional datasets.

I've chosen to only make use of this shared locking in a few code paths
where I already observed contention being a problem, but I do believe a
shared lock could be acquired in many functions here. My preference is
to keep the diff small.

nickitat commented on src/Storages/MergeTree/MergeTreeData.cpp:
```diff
@@ -9794,30 +9841,33 @@ StorageSnapshotPtr MergeTreeData::createStorageSnapshot(const StorageMetadataPtr
     /// Avoid holding the lock while constructing RangesInDataParts
-    snapshot_data->parts = RangesInDataParts(parts);
+    if (snapshot_data->parts == nullptr)
+        snapshot_data->parts = std::make_shared<const RangesInDataParts>(RangesInDataParts(*parts));
```
```suggestion
        snapshot_data->parts = std::make_shared<const RangesInDataParts>(*parts);
```

nickitat commented on src/Storages/MergeTree/MergeTreeData.h:
```diff
@@ -93,13 +93,28 @@ namespace ErrorCodes
 
 struct DataPartsLock
 {
+    explicit DataPartsLock(std::shared_mutex & data_parts_mutex_, std::function<void(DataPartsLock&)> callback_);
```
```suggestion
    DataPartsLock(std::shared_mutex & data_parts_mutex_, std::function<void(DataPartsLock&)> callback_);
```

azat commented:
Thanks for sharing!

> As for lockParts vs readLockParts, I think we must call lockParts, since an INSERT is actually modifying the parts list (by adding new parts), so it would not be safe to do that concurrently with another insert or merge which would also be mutating the parts list at the same time.

You are right, I was not suggesting to use `readLockParts` in `INSERT`, but more of a find missing places where we can use `readLockParts` instead, maybe there are some left.

> IMO the extra cost of an INSERT is in making an extra copy of the parts vector, which actually you've already got happening on every single SELECT query currently. So it basically should just be moving that cost from one place to another. If you had a usecase where there's more inserts than selects then I guess it's a bad change, but for us at least we get a lot more selects. Possibly I am missing something else that adds cost too, lmk what you think

I doubt that copying vector will slows INSERTs 5x times, maybe there is something else?

azat commented on src/Processors/QueryPlan/ReadFromMergeTree.cpp:
```diff
@@ -2012,7 +2012,7 @@ ReadFromMergeTree::AnalysisResultPtr ReadFromMergeTree::selectRangesToRead(
 
     {
-        MergeTreeDataSelectExecutor::filterPartsByPartition(
+        auto res = MergeTreeDataSelectExecutor::filterPartsByPartition(
```
Can we leave `parts` passed not by const ref (we pass it as rvalue anyway for SELECTs) and use `parts = ` here?

canhld94 commented:
This is cool, we're seeing same problems at Ahrefs and I'm having same idea to use shared mutex for in-memory parts set index.

azat reviewed (commented):
@jawm can you PTAL at the coding style issues (and fix it everywhere in the code), merge with master (to fix conflicts) and in the mean time I will review it.

azat committed:
Accept DataPartsAnyLock for others getActiveContainingPart() (required for private)

azat committed:
Data parts lock split review

azat reviewed (approved):
Thanks! Mostly looks ok, let me finish it (all review comments are just for information, I will update everything by myself), I will push couple of commits on top and also update the private part.

_And please don't rebase the original branch, otherwise conflict resolution with private will be lost_

azat committed:
Fix readability-inconsistent-declaration-parameter-name for getVisibleDataPartsVectorUnlocked()

azat committed:
Remove adjustDataPartsVectorBasedOnCacheWarmness() (private method)

azat committed:
Allow any lock in MergeTreeData::grabActivePartsToRemoveForDropRange()

azat committed:
Merge remote-tracking branch 'u/master' into jawm-readlock-pull

azat committed:
Merge remote-tracking branch 'u/master' into jawm-readlock-pull

azat committed:
Replace some lockParts() with readLockParts()

azat commented:
>i'm guessing you meant lockParts?

Yep.

I meant that in some places we acquire this lock twice, which means that we will update list of parts multiple times, while before it was zero. This may or may not be a big deal. I will review the code.

azat reviewed (approved):
Now need to wait for perf tests (there is pending fix for them)

Rest of CI:
- `test_ddl_worker_with_loopback_hosts` - https://github.com/ClickHouse/ClickHouse/pull/88472
- https://github.com/ClickHouse/ClickHouse/issues/87887
- ~~https://github.com/ClickHouse/ClickHouse/issues/88565~~
- Perf tests - https://github.com/ClickHouse/ClickHouse/issues/88539

azat committed:
tests: disable test_cancel_freeze under sanitizers

CI: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=85535&sha=latest&name_0=PR&name_1=Integration+tests+%28amd_tsan%2C+5%2F6%29

azat committed:
tests: ignore if shadow directory does not exist in test_cancel_freeze

In case of test is skipped under sanitizer it is the case

azat committed:
Merge remote-tracking branch 'u/master' into jawm-readlock-pull

azat committed:
Merge remote-tracking branch 'u/master' into jawm-readlock-pull

azat committed:
Merge remote-tracking branch 'u/master' into jawm-readlock-pull

azat committed:
Merge remote-tracking branch 'u/master' into jawm-readlock-pull

azat committed:
More profile events

azat committed:
Update shared parts/ranges on SELECT not on INSERT

v2: Use parts lock for protecting shared parts/ranges and reset them on write lock acquire

azat committed:
Avoid extra copying of RangesInDataParts

azat committed:
Merge remote-tracking branch 'u/master' into dev

azat committed:
Fix tidy warning in MergeTreeDataSelectExecutor

azat committed:
Fix some variable names in MergeTreeDataSelectExecutor

azat committed:
Return order of args for preparePartForCommit()

azat committed:
Update some comments

azat committed:
Polish MergeTreeData::getPossiblySharedVisibleDataPartsRanges()

azat committed:
Avoid holding data parts lock during loading part in MergeTreeData::refreshDataParts()

azat committed:
Fix typo

azat committed:
Fix race in MergeTreeData::getPossiblySharedVisibleDataPartsRanges()

azat committed:
Revert changes in tests since shared parts updated on SELECT now

azat committed:
Fix obtaining parts for SELECT in transactions

azat commented:
So:
- The INSERT performance restored (now shared parts updated on first SELECT, and each write lock resets it)
- Redundant copy of `RangesInDataParts` has been removed which gives 3.5x performance benefit with 100K parts for simple SELECT (when partition pruning removes everything except one part)

And it is ready to land

azat commented on src/Storages/MergeTree/MergeTreeData.cpp:
```diff
@@ -6993,6 +7034,47 @@ DataPartsVector MergeTreeData::getVisibleDataPartsVector(ContextPtr local_contex
+
+            /// Convert read to write lock, and re-check the shared_parts_list
+            shared_lock_holder.reset();
+            lock_holder.emplace(data_parts_mutex, /*data_=*/ nullptr);
```
Comment is useless, but the place is correct - https://github.com/ClickHouse/ClickHouse/pull/93022
